realm/realm-library/src/main/java/io/realm/Realm.java

```diff
@@ -1323,14 +1255,14 @@ public void run() {
 
                         if (backgroundException != null) {
                             if (onError != null) {
-                                handler.post(new Runnable() {
+                                handlerController.asyncTransactionCallbacks.add(new Runnable() {
```

Comment:
If there is an error, no need to wait for REALM_CHANGED, it can be posted to the handler right away.